File: problems/035-wait-for-cycles/solutions/python/main.py
import sys

a = list(map(int, sys.stdin.read().split()))
n, m = a[:2]
g = [[] for _ in range(n)]
rg = [[] for _ in range(n)]
edges = []
selfloop = [False] * n
p = 2
for _ in range(m):
    u, v = a[p] - 1, a[p + 1] - 1
    p += 2
    g[u].append(v)
    rg[v].append(u)
    edges.append((u, v))
    selfloop[u] |= u == v
seen = [False] * n
cur = [0] * n
order = []
for s in range(n):
    if seen[s]:
        continue
    seen[s] = True
    st = [s]
    while st:
        u = st[-1]
        if cur[u] < len(g[u]):
            v = g[u][cur[u]]
            cur[u] += 1
            if not seen[v]:
                seen[v] = True
                st.append(v)
        else:
            order.append(u)
            st.pop()
comp = [-1] * n
members = []
for s in reversed(order):
    if comp[s] >= 0:
        continue
    cid = len(members)
    mem = []
    comp[s] = cid
    st = [s]
    while st:
        u = st.pop()
        mem.append(u)
        for v in rg[u]:
            if comp[v] < 0:
                comp[v] = cid
                st.append(v)
    members.append(mem)
members = [[] for _ in members]
for i, c in enumerate(comp):
    members[c].append(i)
indeg = [False] * len(members)
for u, v in edges:
    if comp[u] != comp[v]:
        indeg[comp[v]] = True
cyc = []
for i in range(n):
    c = comp[i]
    if members[c][0] == i and (len(members[c]) > 1 or selfloop[i]):
        cyc.append(members[c])
out = [f"{len(cyc)} {sum(not x for x in indeg)}"]
out += [str(len(x)) + " " + " ".join(str(v + 1) for v in x) for x in cyc]
print("\n".join(out))

File: problems/035-wait-for-cycles/solutions/rust/main.rs
use std::fmt::Write;
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut it = s.split_whitespace();
    let n: usize = it.next().unwrap().parse().unwrap();
    let m: usize = it.next().unwrap().parse().unwrap();
    let (mut g, mut rg) = (vec![vec![]; n], vec![vec![]; n]);
    let mut edges = vec![];
    let mut selfloop = vec![false; n];
    for _ in 0..m {
        let u = it.next().unwrap().parse::<usize>().unwrap() - 1;
        let v = it.next().unwrap().parse::<usize>().unwrap() - 1;
        g[u].push(v);
        rg[v].push(u);
        edges.push((u, v));
        selfloop[u] |= u == v;
    }
    let mut seen = vec![false; n];
    let mut cur = vec![0; n];
    let mut order = vec![];
    for root in 0..n {
        if seen[root] {
            continue;
        }
        seen[root] = true;
        let mut st = vec![root];
        while let Some(&u) = st.last() {
            if cur[u] < g[u].len() {
                let v = g[u][cur[u]];
                cur[u] += 1;
                if !seen[v] {
                    seen[v] = true;
                    st.push(v)
                }
            } else {
                order.push(u);
                st.pop();
            }
        }
    }
    let mut comp = vec![usize::MAX; n];
    let mut cc = 0;
    for &root in order.iter().rev() {
        if comp[root] != usize::MAX {
            continue;
        }
        comp[root] = cc;
        let mut st = vec![root];
        while let Some(u) = st.pop() {
            for &v in &rg[u] {
                if comp[v] == usize::MAX {
                    comp[v] = cc;
                    st.push(v)
                }
            }
        }
        cc += 1
    }
    let mut mem = vec![vec![]; cc];
    for i in 0..n {
        mem[comp[i]].push(i)
    }
    let mut indeg = vec![false; cc];
    for &(u, v) in &edges {
        if comp[u] != comp[v] {
            indeg[comp[v]] = true
        }
    }
    let wake = indeg.iter().filter(|&&x| !x).count();
    let mut cyc = vec![];
    for i in 0..n {
        let c = comp[i];
        if mem[c][0] == i && (mem[c].len() > 1 || selfloop[i]) {
            cyc.push(c)
        }
    }
    let mut out = format!("{} {}\n", cyc.len(), wake);
    for c in cyc {
        write!(out, "{}", mem[c].len()).unwrap();
        for &v in &mem[c] {
            write!(out, " {}", v + 1).unwrap()
        }
        out.push('\n');
    }
    print!("{out}");
}
